p2p/raintree/network.go

```diff
@@ -25,49 +27,61 @@ var _ typesP2P.Network = &rainTreeNetwork{}
 type rainTreeNetwork struct {
 	base_modules.IntegratableModule
 
-	selfAddr       cryptoPocket.Address
-	pstoreProvider peerstore_provider.PeerstoreProvider
-
-	peersManager *rainTreePeersManager
-	nonceDeduper *mempool.GenericFIFOSet[uint64, uint64]
-
-	currentHeightProvider providers.CurrentHeightProvider
-
 	logger *modules.Logger
+	// host represents a libp2p network node, it encapsulates a libp2p peerstore
+	// & connection manager. `libp2p.New` configures and starts listening
+	// according to options.
+	// (see: https://pkg.go.dev/github.com/libp2p/go-libp2p#section-readme)
+	host libp2pHost.Host
+	// selfAddr is the pocket address representing this host.
+	selfAddr              cryptoPocket.Address
+	peersManager          *rainTreePeersManager
+	pstoreProvider        peerstore_provider.PeerstoreProvider
+	currentHeightProvider providers.CurrentHeightProvider
+	nonceDeduper          *mempool.GenericFIFOSet[uint64, uint64]
 }
 
-func NewRainTreeNetwork(addr cryptoPocket.Address, bus modules.Bus, pstoreProvider providers.PeerstoreProvider, currentHeightProvider providers.CurrentHeightProvider) typesP2P.Network {
+// TECHDEBT: refactor signature to receive a config struct
```

Comment:
NIT: Maybe suggestion to follow the `Create` factory pattern we already use?

Reply:
Great suggestion - thank you! :raised_hands: 

My apologies if I got a bit carried away (i.e. creating extra code to review) but I went ahead with the config struct refactor at the same time. :sweat_smile: As I was thinking about it though I had an idea about how we might formally define this such that we can enforce it - here's where I got to (will include in a PR at some point):

```golang
// e.g. shared/modules/module.go
type FactoryWithOptions[T interface{}] interface {
	Create(bus Bus, opts ...ModuleOption) (T, error)
}

type FactoryWithRequired[T interface{}, K interface{}] interface {
	Create(bus Bus, cfg K) (T, error)
}

type FactoryWithRequiredAndOptions[T interface{}, K interface{}] interface {
	Create(bus Bus, cfg K, opts ...ModuleOption) (T, error)
}
```

```golang
// e.g. p2p/raintree/network.go
var _ modules.FactoryWithRequired[typesP2P.Network, RainTreeConfig] = &rainTreeNetwork{}
```